src/praisonai/praisonai/bots/_session.py

```
@@ -384,6 +384,56 @@ def _save_history(
+        """
+        if tool_policy is None:
+            return None
+        filter_tools = getattr(tool_policy, "filter_tools", None)
+        if not callable(filter_tools):
+            return None
+        original = getattr(agent, "tools", None)
+        if not original:
+            return None
+        try:
+            filtered = filter_tools(list(original))
+        except Exception as e:  # pragma: no cover - defensive
```

Ruff (0.15.18)

[warning] Do not catch blind exception: `Exception`

(BLE001)